api/src/schedules/retention.ts

```diff
@@ -0,0 +1,115 @@
+import { Action } from '@directus/constants';
+import { useEnv } from '@directus/env';
+import { toBoolean } from '@directus/utils';
+import type { Knex } from 'knex';
+import getDatabase from '../database/index.js';
+import { useLock } from '../lock/index.js';
+import { useLogger } from '../logger/index.js';
+import { getMilliseconds } from '../utils/get-milliseconds.js';
+import { scheduleSynchronizedJob, validateCron } from '../utils/schedule.js';
+
+export interface RetentionTask {
+	collection: string;
+	where?: readonly [string, string, Knex.Value | null];
+	join?: readonly [string, string, string];
+	timeframe?: number;
+}
+
+const env = useEnv();
+
+const retentionLockKey = 'schedule--data-retention';
+const retentionLockTimeout = 10 * 60 * 1000; // 10 mins
+
+const RETENTION_TASKS: RetentionTask[] = [
+	{
+		collection: 'directus_activity',
+		where: ['action', '!=', Action.RUN],
+		timeframe: getMilliseconds(env['ACTIVITY_RETENTION']),
+	},
+	{
+		collection: 'directus_activity',
+		where: ['action', '=', Action.RUN],
+		timeframe: getMilliseconds(env['FLOW_LOGS_RETENTION']),
+	},
+	{
+		collection: 'directus_revisions',
+		join: ['directus_activity', 'directus_revisions.activity', 'directus_activity.id'],
+		timeframe: getMilliseconds(env['REVISIONS_RETENTION']),
+	},
```

Comment:
Should this task for `REVISIONS_RETENTION` be skipped if the `ACTIVITY_RETENTION` is enabled as the deletions are cascaded? 🤔

Reply:
Only if activity <= revisions retention

Reply:
Resolved via e9919ae53f9bf8983586ec636124da965a6ffb14